### ISSUE TITLE
Change in how tx data is modeled

### ISSUE THREAD

**Issue by Pongch** [omisego/elixir-omg#844]
NET-71

As a consumer of abstract layer contracts (operator and watcher), I would like to remodel my data to support new tx types, So that I could have future-proof tx upgrades in the future

**Risk**

- big refactor with many breaking changes
- potential for breaking changes in the future (mass exits incompatibility)

**Value**
- Future tx type support
- avoiding hard upgrade process

### Implementation

A rough breakdown of changes & progress is this:

 - [x] refactor child chain (`OMG.State.Core.exec` & `Transaction.Recovered.recover_from` flows), to be able to add new tx types easily #882
 - [x] experiment with settlement tx to validate the above
 - [x] refactor child chain (`OMG.State.Core.exec` & `Transaction.Recovered.recover_from` flows) to be able to add new output types easily #1036 
 - [x] experiment with ERC721 tx to validate the above (just a dirty [branch here](https://github.com/omisego/elixir-omg/tree/experimental/844_partial_erc721_support) for now; waiting for demand of a tidier integration/demo)
 - [x] experiment with MVP-based txs to validate the above ([branch here](https://github.com/omisego/elixir-omg/tree/844_experimental_tx_type_abstract_outputs), slightly outdated now, but seemed to hold water + require a bit of futher steps)
 - [x] refactor child chain to be able to modify/add input pointer types easily #1036 
 - [x] experiment with `output_id`-driven txs to validate the above ([branch here](https://github.com/omisego/elixir-omg/tree/844_draft_output_id), slightly outdated now, seemed to hold water)
 - [x] refactor watcher-related parts (double spend detection and others) to be able to add new tx types easily (probably still would require some effort on the informational api/postgres end)
 - [ ] ~experiment with different tx types to validate the above~ I'd opt to postpone this

**Comment by pdobacz** [comment]
left some notes in the checkbox-list in the PR description

**Comment by pdobacz** [comment]
@Pongch I'm revisiting this from the user-story PoV and wondering what would the acceptance criteria be here.

A lot has been discussed on this topic and still there's a couple of open points that could pivot the data model, e.g.:
 - MVP/MoreVP
 - some dex-stuff pending
 - approach to informational APIs on the Watcher
 - handling multiple ExitGames (which I think pertains more to the other ALD issues)

So far, what we got to is a lower the barrier of entry for new tx types with respect to the data model. To update the core functionality by introducing a new tx type, one would need to:
 - implement a specified APIs for new tx type (and possibly output type/input pointer type)
 - point to those APIs in the services configuration, to allow child chain/watcher to acknowledge the new tx

To maybe better describe the result: it should be rather straightforward to introduce a new MoreVP-based tx type, insofar as to be able to deposit/transact/standard-exit in the chch/watcher and have those services ingest the tx. IFEs, MVP, informational APIs and contract management is still a bit loose.

(I'll invest some time now to possibly have a quick :shell: demo of ERC721 support, to show what I mean).

In a nutshell, not sure what the fate of this ticket be now :thinking:

**Comment by pdobacz** [comment]
A gentle prod @Pongch on the above question: what should we do with this ticket here? Are the criteria satisfied? How should we move forward?

**Comment by pdobacz** [comment]
@Pongch ping again, good to close? /cc @achiurizo

**Comment by InoMurko** [comment]
This seems to be done. Re-open if it's not @achiurizo @Pongch 

Tnx @pdobacz 👯‍♂
